sdk/core/allocator/main.cc

```diff
@@ -863,22 +863,54 @@ __cheriot_minimum_stack(0x90) ssize_t
 	return cap->quota;
 }
 
-__cheriot_minimum_stack(0xd0) int heap_quarantine_empty()
+__cheriot_minimum_stack(0xd0) int heap_quarantine_flush(Timeout *timeout)
 {
 	STACK_CHECK(0xd0);
-	LockGuard g{lock};
-	while (gm->heapQuarantineSize > 0)
+
+	if (!check_timeout_pointer(timeout))
+	{
+		return -EINVAL;
+	}
+
+	if (LockGuard g{lock, timeout})
 	{
-		if (!gm->quarantine_dequeue())
+		while (gm->heapQuarantineSize > 0)
 		{
-			revoker.system_bg_revoker_kick();
+			auto epoch = revoker.system_epoch_get();
+			if (!gm->quarantine_dequeue())
+			{
+				revoker.system_bg_revoker_kick();
+			}
+			if constexpr (Revocation::Revoker::IsAsynchronous)
+			{
+				wait_for_background_revoker(timeout, epoch, g);
+			}
+			else
+			{
+				(void)epoch;
+				// Drop and reacquire the lock while yielding.
```

Comment:
This sequence could be extracted as a method on `LockGuard` so that this would become `if (!g.unlock_temporarily(timeout, 1)) return -ETIMEDOUT;`. Do we have other instances of this pattern?